File: Sources/AtticCore/ServerLauncher.swift
// =============================================================================
// ServerLauncher.swift - AtticServer Auto-Launch Utility
// =============================================================================
//
// This file provides a shared utility for launching AtticServer as a subprocess.
// It is used by AtticCLI, AtticGUI, and AtticMCP to auto-start the server when
// one isn't already running.
//
// Usage:
//   let launcher = ServerLauncher()
//   if let socketPath = launcher.launchServer(silent: true) {
//       // Connect to socketPath
//   }
//
// The launcher searches for AtticServer in several locations:
// 1. Same directory as the current executable
// 2. PATH environment variable
// 3. Common installation locations (/usr/local/bin, /opt/homebrew/bin, etc.)
//
// =============================================================================

import Foundation

// MARK: - Server Launcher

/// Configuration options for launching AtticServer.
public struct ServerLaunchOptions: Sendable {
    /// Whether to disable audio output.
    public var silent: Bool

    /// Custom ROM path (optional).
    public var romPath: String?

    /// Whether to suppress stdout/stderr from the server process.
    public var suppressOutput: Bool

    /// Creates launch options with defaults.
    public init(
        silent: Bool = false,
        romPath: String? = nil,
        suppressOutput: Bool = true
    ) {
        self.silent = silent
        self.romPath = romPath
        self.suppressOutput = suppressOutput
    }
}

/// Result of a server launch attempt.
public enum ServerLaunchResult: Sendable {
    /// Server launched successfully.
    /// - socketPath: Path to the CLI socket.
    /// - pid: Process ID of the server.
    case success(socketPath: String, pid: Int32)

    /// Server executable not found.
    case executableNotFound

    /// Failed to start the process.
    case launchFailed(Error)

    /// Server started but socket didn't appear in time.
    case socketTimeout(pid: Int32)
}

/// Utility for finding and launching AtticServer.
///
/// This class provides methods to:
/// - Find an existing running AtticServer
/// - Launch a new AtticServer instance
/// - Wait for the server to be ready
///
/// Example usage:
/// ```swift
/// let launcher = ServerLauncher()
///
/// // Try to find existing server first
/// if let socketPath = launcher.discoverExistingServer() {
///     print("Found existing server at \(socketPath)")
/// } else {
///     // Launch new server
///     switch launcher.launchServer() {
///     case .success(let socketPath, let pid):
///         print("Launched server (PID: \(pid)) at \(socketPath)")
///     case .executableNotFound:
///         print("AtticServer not found")
///     case .launchFailed(let error):
///         print("Launch failed: \(error)")
///     case .socketTimeout(let pid):
///         print("Server started but socket not ready")
///     }
/// }
/// ```
public final class ServerLauncher: Sendable {

    // =========================================================================
    // MARK: - Initialization
    // =========================================================================

    public init() {}

    // =========================================================================
    // MARK: - Server Discovery
    // =========================================================================

    /// Discovers an existing running AtticServer.
    ///
    /// Scans `/tmp/attic-*.sock` for available sockets.
    ///
    /// - Returns: Path to an existing socket, or nil if none found.
    public func discoverExistingServer() -> String? {
        let client = CLISocketClient()
        return client.discoverSocket()
    }

    // =========================================================================
    // MARK: - Server Launch
    // =========================================================================

    /// Launches AtticServer as a subprocess.
    ///
    /// This method:
    /// 1. Finds the AtticServer executable
    /// 2. Launches it with the specified options
    /// 3. Waits for the CLI socket to appear
    ///
    /// - Parameter options: Launch configuration options.
    /// - Returns: The launch result indicating success or failure mode.
    public func launchServer(options: ServerLaunchOptions = ServerLaunchOptions()) -> ServerLaunchResult {
        // Find the server executable
        guard let serverPath = findServerExecutable() else {
            return .executableNotFound
        }

        // Build arguments
        var arguments: [String] = []
        if options.silent {
            arguments.append("--silent")
        }
        if let romPath = options.romPath {
            arguments.append("--rom-path")
            arguments.append(romPath)
        }

        // Create and configure the process
        let process = Process()
        process.executableURL = URL(fileURLWithPath: serverPath)
        process.arguments = arguments

        // Suppress output if requested
        if options.suppressOutput {
            process.standardOutput = FileHandle.nullDevice
            process.standardError = FileHandle.nullDevice
        }

        // Launch the process
        do {
            try process.run()
        } catch {
            return .launchFailed(error)
        }

        let pid = process.processIdentifier
        let socketPath = CLIProtocolConstants.socketPath(for: pid)

        // Wait for socket to appear (with timeout)
        let maxRetries = 20  // 4 seconds total
        let retryInterval: TimeInterval = 0.2

        for _ in 0..<maxRetries {
            if FileManager.default.fileExists(atPath: socketPath) {
                return .success(socketPath: socketPath, pid: pid)
            }
            Thread.sleep(forTimeInterval: retryInterval)
        }

        return .socketTimeout(pid: pid)
    }

    /// Convenience method that discovers or launches a server.
    ///
    /// First tries to find an existing server. If none found, launches a new one.
    ///
    /// - Parameter options: Launch options if a new server needs to be started.
    /// - Returns: Socket path on success, nil on failure.
    public func ensureServerRunning(options: ServerLaunchOptions = ServerLaunchOptions()) -> String? {
        // Try to find existing server first
        if let existingSocket = discoverExistingServer() {
            return existingSocket
        }

        // Launch new server
        switch launchServer(options: options) {
        case .success(let socketPath, _):
            return socketPath
        case .executableNotFound, .launchFailed, .socketTimeout:
            return nil
        }
    }

    // =========================================================================
    // MARK: - Executable Discovery
    // =========================================================================

    /// Finds the AtticServer executable.
    ///
    /// Searches in several locations:
    /// 1. Same directory as the current executable
    /// 2. PATH environment variable
    /// 3. Common installation locations
    ///
    /// - Returns: Path to AtticServer executable, or nil if not found.
    public func findServerExecutable() -> String? {
        let fileManager = FileManager.default

        // Get the directory containing the current executable
        let executablePath = CommandLine.arguments[0]
        let executableDir = (executablePath as NSString).deletingLastPathComponent

        // Check in same directory as current executable
        let sameDirPath = (executableDir as NSString).appendingPathComponent("AtticServer")
        if fileManager.isExecutableFile(atPath: sameDirPath) {
            return sameDirPath
        }

        // Check in PATH environment variable
        if let pathEnv = ProcessInfo.processInfo.environment["PATH"] {
            for dir in pathEnv.split(separator: ":") {
                let path = (String(dir) as NSString).appendingPathComponent("AtticServer")
                if fileManager.isExecutableFile(atPath: path) {
                    return path
                }
            }
        }

        // Check common installation locations
        let commonPaths = [
            "/usr/local/bin/AtticServer",
            "/opt/homebrew/bin/AtticServer",
            "~/.local/bin/AtticServer"
        ]

        for path in commonPaths {
            let expandedPath = (path as NSString).expandingTildeInPath
            if fileManager.isExecutableFile(atPath: expandedPath) {
                return expandedPath
            }
        }

        return nil
    }
}

File: Python/AtticCLI/attic_cli/server_launcher.py
"""Find and launch AtticServer subprocess.

Ported from Sources/AtticCore/ServerLauncher.swift. Searches for the
AtticServer executable in standard locations and launches it, waiting
for the socket to appear.
"""

import logging
import os
import shutil
import subprocess
import sys
import time
from dataclasses import dataclass
from pathlib import Path

from .protocol import socket_path_for_pid

logger = logging.getLogger(__name__)

# How long to wait for the server socket to appear after launch.
_POLL_INTERVAL = 0.2  # seconds
_POLL_MAX_RETRIES = 20  # 20 * 0.2 = 4 seconds total


@dataclass(frozen=True, slots=True)
class LaunchResult:
    """Result of attempting to launch AtticServer.

    Attributes:
        success: True if the server is running and the socket is available.
        socket_path: Path to the Unix domain socket (if successful).
        pid: Server process ID (if launched).
        error: Error description (if failed).
    """

    success: bool
    socket_path: str = ""
    pid: int = 0
    error: str = ""


def find_server_executable() -> str | None:
    """Search for the AtticServer executable.

    Checks (in order):
    1. Same directory as the current executable (co-located builds)
    2. .build/release/ and .build/debug/ in the project root
    3. PATH via shutil.which()
    4. Common installation directories

    Returns:
        Absolute path to AtticServer, or None if not found.
    """
    name = "AtticServer"

    # 1. Same directory as the current executable
    exe_dir = Path(sys.argv[0]).resolve().parent
    candidate = exe_dir / name
    if candidate.is_file() and os.access(candidate, os.X_OK):
        return str(candidate)

    # 2. Swift build directories (relative to project root)
    # Walk up from the package directory to find the project root
    pkg_dir = Path(__file__).resolve().parent.parent  # attic_cli -> AtticCLI
    project_root = pkg_dir.parent.parent  # AtticCLI -> Python -> attic
    for build_dir in ["release", "debug"]:
        candidate = project_root / ".build" / build_dir / name
        if candidate.is_file() and os.access(candidate, os.X_OK):
            return str(candidate)

    # 3. PATH
    found = shutil.which(name)
    if found:
        return found

    # 4. Common installation directories
    common_dirs = [
        "/usr/local/bin",
        "/opt/homebrew/bin",
        os.path.expanduser("~/.local/bin"),
    ]
    for d in common_dirs:
        candidate = Path(d) / name
        if candidate.is_file() and os.access(candidate, os.X_OK):
            return str(candidate)

    return None


def launch_server(
    *,
    silent: bool = False,
    rom_path: str | None = None,
) -> LaunchResult:
    """Launch AtticServer and wait for its socket to appear.

    Args:
        silent: Pass --silent to suppress audio output.
        rom_path: Pass --rom-path to specify ROM directory.

    Returns:
        LaunchResult indicating success or failure.
    """
    exe = find_server_executable()
    if exe is None:
        return LaunchResult(
            success=False,
            error=(
                "AtticServer executable not found. "
                "Build it with 'swift build' or ensure it's on your PATH."
            ),
        )

    args = [exe]
    if silent:
        args.append("--silent")
    if rom_path:
        args.extend(["--rom-path", rom_path])

    logger.debug("Launching AtticServer: %s", " ".join(args))

    try:
        proc = subprocess.Popen(
            args,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )
    except OSError as exc:
        return LaunchResult(success=False, error=f"Failed to launch AtticServer: {exc}")

    pid = proc.pid
    sock_path = socket_path_for_pid(pid)

    # Poll for the socket to appear
    for _ in range(_POLL_MAX_RETRIES):
        time.sleep(_POLL_INTERVAL)
        if os.path.exists(sock_path):
            logger.debug("AtticServer socket ready: %s (pid %d)", sock_path, pid)
            return LaunchResult(success=True, socket_path=sock_path, pid=pid)

    return LaunchResult(
        success=False,
        pid=pid,
        error=f"AtticServer launched (pid {pid}) but socket did not appear within 4 seconds.",
    )


def ensure_server_running(
    *,
    silent: bool = False,
    rom_path: str | None = None,
    discover_fn=None,
) -> LaunchResult:
    """Ensure AtticServer is running, launching it if necessary.

    Args:
        silent: Pass --silent when launching a new server.
        rom_path: Pass --rom-path when launching a new server.
        discover_fn: Optional callable that returns an existing socket path.
            Defaults to CLISocketClient.discover_socket behavior.

    Returns:
        LaunchResult for the running server.
    """
    # Try to find an existing server first
    if discover_fn is not None:
        existing = discover_fn()
    else:
        from .cli_client import CLISocketClient

        existing = CLISocketClient().discover_socket()

    if existing is not None:
        # Extract PID from the socket path
        basename = os.path.basename(existing)
        pid_str = basename.removeprefix("attic-").removesuffix(".sock")
        try:
            pid = int(pid_str)
        except ValueError:
            pid = 0
        return LaunchResult(success=True, socket_path=existing, pid=pid)

    # No existing server — launch one
    return launch_server(silent=silent, rom_path=rom_path)
